plugins/micromega/coq_micromega.ml

```diff
@@ -1222,17 +1233,31 @@ let prodn n env b =
   in
   prodrec (n,env,b)
 
+(** [make_goal_of_formula depxr vars props form] where
+     - vars is an environment for the arithmetic variables occurring in form
+     - props is an environment for the propositions occurring in form
+    @return a goal where all the variables and propositions of the formula are quantified
+
+*)
+
 let make_goal_of_formula gl dexpr form =
 
   let vars_idx =
     List.mapi (fun i v -> (v, i+1))  (ISet.elements (var_env_of_formula form)) in
 
   (*  List.iter (fun (v,i) -> Printf.fprintf stdout "var %i has index %i\n" v i) vars_idx ;*)
-  
+
   let props     =  prop_env_of_formula gl form in
 
-  let vars_n  = List.map (fun (_,i) -> (Names.Id.of_string (Printf.sprintf "__x%i" i)) , dexpr.interp_typ) vars_idx in 
-  let props_n = List.mapi (fun i _ -> (Names.Id.of_string (Printf.sprintf "__p%i" (i+1))) , EConstr.mkProp)  (Env.elements props) in
+  (* Not quite fresh... but fast and fresh enough *)
+  let fresh_var str i = Names.Id.of_string (str^(string_of_int i)) in
+
+  let fresh_prop str i =
+    (* TODO: Should also ensure freshness! *)
+    Names.Id.of_string (str^(string_of_int i)) in
```

Comment:
What can go wrong if the generated names are not fresh?

Reply:
The tactic fails with "Error: Variable __pxxx is already declared.".
That is not too bad.

A solution would be to clear all the hypotheses from the context.
This would better be constant time i.e. unsafe update of the evar_map.

Reply:
This is actually fixed by 106087e